crates/vm/wasm/src/vm.rs

```diff
@@ -147,6 +129,67 @@ impl Vm for WasmVm {
             fe,
         })
     }
+
+    fn update_pinned(&self, storage: &dyn Storage) -> Result<(), Self::Error> {
+        let mut new_pinned = match CONFIG.load(storage) {
+            Ok(config) => config.pinned_hashes,
+            // Chain is not init yet
+            Err(_) => return Ok(()),
+        };
+
+        self.cache
+            .pinned
+            .write_with(|mut currently_pinned| -> Result<_, Self::Error> {
+                for i in currently_pinned.keys().cloned().collect::<Vec<_>>() {
+                    if new_pinned.contains(&i) {
+                        new_pinned.remove(&i);
+                    } else {
+                        // Remove from cache if not pinned anymore
+                        currently_pinned.remove(&i);
+                    }
+                }
+
+                for i in new_pinned {
+                    // Remove from default cache
+                    self.cache.try_remove_from_modules(&i);
+
+                    let code = CODES.load(storage, &i)?;
+                    // Signlepass still here, need to be changed with LLVM
+                    let mut compiler = Singlepass::new();
+                    compiler.canonicalize_nans(true);
+                    let (module, engine) = build_from_compiler(compiler, &code)?;
+                    currently_pinned.insert(i, (module, engine));
+                }
+
+                Ok(())
+            })?;
+
+        Ok(())
+    }
+}
+
+pub fn build_from_compiler<C: CompilerConfig + Into<Engine>>(
+    mut compiler: C,
+    code: &[u8],
+) -> VmResult<(Module, Engine)> {
```

Comment:
I suggested using the `with` syntax for all trait bounds throughout our codebase, for consistency. It'a bit more verbose, but more readable imo.

```rust
fn build_from_compiler<C>(mut compiler: C, code: &[u8]) -> VmResult<(Module, Engine)>
where
    C: CompilerConfig + Into<Engine>,
{
    // ...
}
```

Also, does this function needs to be public?